vamb/reclustering.py

```diff
@@ -0,0 +1,319 @@
+    if min_contig_len is not None:
+        contig_len = {h:len(seq) for h,seq in fasta_iter(fasta_path)}
+        data = data[data['contig'].map(lambda c: contig_len[c] >= min_contig_len)]
+    data = data.drop_duplicates(['gene', 'contig'])
+
+    if contig_to_marker:
+        from collections import defaultdict
+        marker = data['gene'].values
+        contig = data['contig'].str.split('.').str[-1].values
+        sequence2markers = defaultdict(list)
+        for m, c in zip(marker, contig):
+            sequence2markers[c].append(m)
+        return sequence2markers
+    else:
+        def extract_seeds(vs, sel):
+            vs = vs.sort_values()
+            median = vs[len(vs) //2]
+
+            # the original version broke ties by picking the shortest query, so we
+            # replicate that here:
+            candidates = vs.index[vs == median]
+            c = qlen.loc[candidates].idxmin()
+            r = list(sel.query('gene == @c')['contig'])
+            r.sort()
+            return r
+
+
+        if multi_mode:
+            data['bin'] = data['orf'].str.split(pat='.', n=0, expand=True)[0]
+            counts = data.groupby(['bin', 'gene'])['orf'].count()
+            res = {}
+            for b,vs in counts.groupby(level=0):
+                cs = extract_seeds(vs.droplevel(0), data.query('bin == @b', local_dict={'b':b}))
+                res[b] = [c.split('.',1)[1] for c in cs]
+            return res
+        else:
+            counts = data.groupby('gene')['orf'].count()
+            return extract_seeds(counts, data)
+
+
+def run_prodigal(fasta_path, num_process, output):
+    contigs = {}
+    for h, seq in fasta_iter(fasta_path):
+        contigs[h] = seq
+
+    total_len = sum(len(s) for s in contigs.values())
+    split_len = total_len // num_process
+
+    cur = split_len + 1
+    next_ix = 0
+    out = None
+    with contextlib.ExitStack() as stack:
+        for h,seq in contigs.items():
+            if cur > split_len and next_ix < num_process:
+                if out is not None:
+                    out.close()
+                out = open(os.path.join(output, 'contig_{}.fa'.format(next_ix)), 'wt')
+                out = stack.enter_context(out)
+
+                cur = 0
+                next_ix += 1
+            out.write(f'>{h}\n{seq}\n')
+            cur += len(seq)
+
+    try:
+        process = []
+        for index in range(next_ix):
+            with open(os.path.join(output, f'contig_{index}_log.txt') + '.out', 'w') as prodigal_out_log:
+                p = subprocess.Popen(
+                        ['prodigal',
+                         '-i', os.path.join(output, f'contig_{index}.fa'),
+                         '-p', 'meta',
+                         '-q',
+                         '-m', # See https://github.com/BigDataBiology/SemiBin/issues/87
+                         '-a', os.path.join(output, f'contig_{index}.faa')
+                         ],
+                    stdout=prodigal_out_log)
```

Comment:
Let's use Pyrodigal instead. It's a Python/Cython re-implementation of Prodigal that is faster and better maintained. Better, it it has built-in Python bindings, which means you don't have to spawn a subprocess for it.

Reply:
sounds cool!